File: test/OAuthGitHub.Api.Unit.Test/Mocks/SignUpMediatorMock.cs
using System;
using System.Linq.Expressions;
using System.Threading;
using System.Threading.Tasks;
using MediatR;
using Moq;
using OAuthGitHub.Api.Unit.Test.Stubs;
using OAuthGithub.Core.Application;

namespace OAuthGitHub.Api.Unit.Test.Mocks
{
    public static class SignUpMediatorMock
    {
        private static readonly Mock<IMediator> Mock = new();

        private static readonly Expression<Func<IMediator, Task<string>>> Expression =
            mediator => mediator.Send(It.IsAny<RegisterCommand>(), new CancellationToken());

        public static IMediator Mediator()
        {
            Mock.Setup(Expression).ReturnsAsync(TokenStub.Token);
            return Mock.Object;
        }

        public static void MockShouldBeCalled(this IMediator mediator)
        {
            Mock.Verify(Expression, Times.AtLeastOnce);
        }
    }
}

File: test/OAuthGitHub.Api.Unit.Test/Mocks/SignUpMediatorMock.fs
module OAuthGitHub.Api.Unit.Test.Mocks

open System
open System.Linq.Expressions
open System.Threading
open System.Threading.Tasks
open MediatR
open Moq
open OAuthGitHub.Api.Unit.Test.Stubs.Stubs
open OAuthGithub.Core.Application

type Expr =
    static member Quote(e: Expression<Func<IMediator, Task<string>>>) = e

module SignUpMediatorMock =
    let Mock = Mock<IMediator>()

    let Expression =
        Expr.Quote
            (fun mediator -> mediator.Send(It.IsAny<RegisterCommand>(), CancellationToken()))

    let Mediator() =
        let _ = Mock.Setup(Expression).ReturnsAsync(tokenStub)
        Mock.Object

    let HaveBeenCalledMock() =
        Mock.Verify(Expression, Times.AtLeastOnce)
